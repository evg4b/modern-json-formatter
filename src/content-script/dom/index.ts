export * from './helpers';
export * from './elements';
export * from './build-dom';
export * from './build-array-node';
export * from './build-error-node';
export * from './build-object-node';
export * from './build-primitive-nodes';
