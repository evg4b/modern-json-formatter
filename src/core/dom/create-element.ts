export interface Options<K extends keyof HTMLElementTagNameMap = 'span'> {
  element: K;
  content?: string;
  class?: string | string[];
  children?: HTMLElement[] | HTMLElement;
  attributes?: Record<string, string>;
}

export const createElement = <K extends keyof HTMLElementTagNameMap>(options: Options<K>): HTMLElementTagNameMap[K] => {
  const element = document.createElement(options.element);
  if (options.content) {
    element.appendChild(document.createTextNode(options.content));
  }
  if (options.class) {
    const classList = [options.class].flat();
    element.classList.add(...classList);
  }
  if (options.children) {
    const children = [options.children].flat();
    element.append(...children);
  }
  if (options.attributes) {
    Object.entries(options.attributes).forEach(([key, value]) => {
      element.setAttribute(key, value);
    });
  }

  return element;
};
