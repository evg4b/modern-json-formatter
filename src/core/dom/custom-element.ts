export const CustomElement = (selector: string, options?: ElementDefinitionOptions) => {
  return (target: CustomElementConstructor) => {
    customElements.define(selector, target, options);
  };
};
