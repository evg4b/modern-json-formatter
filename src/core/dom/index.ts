export * from './custom-element';
export * from './styled-component';
export * from './create-element';
