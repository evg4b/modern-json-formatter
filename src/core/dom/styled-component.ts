import { registerStyle } from '@core/ui/helpers';

export abstract class StyledComponentElement extends HTMLElement {
  protected readonly shadow = this.attachShadow({ mode: 'closed' });
  protected readonly _styles: HTMLStyleElement;

  protected constructor(styles: string) {
    super();
    this._styles = registerStyle(this.shadow, styles);
  }
}
