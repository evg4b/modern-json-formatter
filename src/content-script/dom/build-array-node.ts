import { ArrayNode } from '@packages/tokenizer';
import { buildNode } from './build-dom';
import { comma, ellipsis, squareBracket, toggle } from './elements';
import { buildInfoNode, element, isValueExpandable } from './helpres';

export const buildArrayNode = ({ items }: ArrayNode) => {
  const arrayNode = element({ class: 'array' });
  arrayNode.appendChild(squareBracket.open());

  if (items.length) {
    const innerNode = element({ class: 'inner' });
    arrayNode.append(innerNode, ellipsis());
    const lastIndex = items.length - 1;
    items.forEach((item, index) => {
      const itemNode = element({ class: 'item' });
      innerNode.appendChild(itemNode);
      if (isValueExpandable(item)) {
        itemNode.appendChild(toggle());
      }
      itemNode.appendChild(buildNode(item));
      if (index !== lastIndex) {
        itemNode.appendChild(comma());
      }
      const infoNode = buildInfoNode(item);
      if (infoNode) {
        itemNode.appendChild(infoNode);
      }
    });
  }
  arrayNode.appendChild(squareBracket.close());

  return arrayNode;
};
