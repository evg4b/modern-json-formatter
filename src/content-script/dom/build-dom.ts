import { createElement } from '@core/dom';
import { TokenNode, TupleNode } from '@packages/tokenizer';
import { buildArrayNode } from './build-array-node';
import { buildObjectNode } from './build-object-node';
import { buildBoolNode, buildNullNode, buildNumberNode, buildStringNode } from './build-primitive-nodes';
import { toggle } from './elements';
import { buildInfoNode, isToggleElement, isValueExpandable } from './helpres';

export const buildDom = (object: TokenNode | TupleNode): HTMLElement => {
  if (object.type === 'tuple') {
    return createElement({
      class: 'tuple',
      element: 'div',
      children: object.items.map(buildDom),
    });
  }

  const root = createElement({ element: 'span', class: 'root' });

  if (isValueExpandable(object)) {
    root.appendChild(toggle());
    root.classList.add('padding');
  }

  root.append(buildNode(object));
  const infoNode = buildInfoNode(object);
  if (infoNode) {
    root.appendChild(infoNode);
  }

  root.addEventListener('click', ({ target }) => {
    if (isToggleElement(target) && target.parentNode instanceof HTMLElement) {
      target.parentNode.classList.toggle('collapsed');
    }
  });

  return root;
};

export const buildNode = (object: TokenNode): HTMLElement => {
  switch (object.type) {
    case 'string':
      return buildStringNode(object);
    case 'number':
      return buildNumberNode(object);
    case 'bool':
      return buildBoolNode(object);
    case 'null':
      return buildNullNode();
    case 'array':
      return buildArrayNode(object);
    case 'object':
      return buildObjectNode(object);
    default:
      throw new Error('Unknown type');
  }
};
