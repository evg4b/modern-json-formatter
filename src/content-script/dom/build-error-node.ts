import { resource } from '@core/browser';
import { createElement } from '@core/dom';

export const buildErrorNode = (header: string, ...lines: string[]) => {
  const wrapper = document.createElement('div');
  wrapper.className = 'error';
  const image = document.createElement('img');

  image.src = resource('invalid.svg');
  image.alt = 'Error';
  const message = document.createElement('div');
  message.className = 'message';
  message.append(
    createElement({ element: 'span', content: header }), // 'Invalid json file.' }),
    ...lines.map(line => createElement({ element: 'span', content: line })),
    // element({ content: 'Please check the file and try again.' }),
  );
  wrapper.append(image, message);

  return wrapper;
};
