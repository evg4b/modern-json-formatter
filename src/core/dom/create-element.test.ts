import { createElement, Options } from '@core/dom/create-element';
import { describe, expect } from '@jest/globals';

describe('createElement', () => {
  const testCases: Options<'span' | 'div' | 'input' | 'a'>[] = [
    { element: 'span' },
    { element: 'span', class: 'test' },
    { element: 'span', class: 'test', content: 'test content' },
    { element: 'span', class: ['test1', 'test2', 'test3'], content: 'test content' },
    {
      element: 'span',
      children: [
        createElement({ element: 'span', content: 'test1' }),
        createElement({ element: 'span', content: 'test2' }),
        createElement({ element: 'span', content: 'test3' }),
      ],
    },
    { element: 'span', content: 'test content', attributes: { ref: 'test' } },
    { element: 'span', html: '<span>test</span>' },
    { element: 'div' },
    { element: 'div', class: 'test' },
    { element: 'div', class: 'test', content: 'test content' },
    { element: 'div', class: ['test1', 'test2', 'test3'], content: 'test content' },
    {
      element: 'div',
      children: [
        createElement({ element: 'div', content: 'test1' }),
        createElement({ element: 'div', content: 'test2' }),
        createElement({ element: 'div', content: 'test3' }),
      ],
    },
    { element: 'div', content: 'test content', attributes: { 'aria-data': 'test' } },
    { element: 'div', html: '<span>test</span>' },
    { element: 'a' },
    { element: 'a', class: 'test' },
    { element: 'a', class: 'test', content: 'test content' },
    { element: 'a', class: ['test1', 'test2', 'test3'], content: 'test content' },
    {
      element: 'a',
      children: [
        createElement({ element: 'span', content: 'test1' }),
        createElement({ element: 'span', content: 'test2' }),
        createElement({ element: 'span', content: 'test3' }),
      ],
    },
    { element: 'a', content: 'test content', attributes: { href: 'http://test' } },
    { element: 'a', html: '<span>test</span>' },
    { element: 'input' },
    { element: 'input', class: 'test' },
    { element: 'input', class: ['test1', 'test2', 'test3'] },
    { element: 'input', content: 'test content', attributes: { test: 'data' } },
  ];

  test.each(testCases)('%p', options => {
    expect(createElement(options)).toMatchSnapshot();
  });
});
