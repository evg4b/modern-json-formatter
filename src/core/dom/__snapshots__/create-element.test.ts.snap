// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createElement {"attributes": [Object], "content": "test content", "element": "a"} 1`] = `
<a
  href="http://test"
>
  test content
</a>
`;

exports[`createElement {"attributes": [Object], "content": "test content", "element": "div"} 1`] = `
<div
  aria-data="test"
>
  test content
</div>
`;

exports[`createElement {"attributes": [Object], "content": "test content", "element": "input"} 1`] = `
<input
  test="data"
>
  test content
</input>
`;

exports[`createElement {"attributes": [Object], "content": "test content", "element": "span"} 1`] = `
<span
  ref="test"
>
  test content
</span>
`;

exports[`createElement {"children": [Array], "element": "a"} 1`] = `
<a>
  <span>
    test1
  </span>
  <span>
    test2
  </span>
  <span>
    test3
  </span>
</a>
`;

exports[`createElement {"children": [Array], "element": "div"} 1`] = `
<div>
  <div>
    test1
  </div>
  <div>
    test2
  </div>
  <div>
    test3
  </div>
</div>
`;

exports[`createElement {"children": [Array], "element": "span"} 1`] = `
<span>
  <span>
    test1
  </span>
  <span>
    test2
  </span>
  <span>
    test3
  </span>
</span>
`;

exports[`createElement {"class": "test", "content": "test content", "element": "a"} 1`] = `
<a
  class="test"
>
  test content
</a>
`;

exports[`createElement {"class": "test", "content": "test content", "element": "div"} 1`] = `
<div
  class="test"
>
  test content
</div>
`;

exports[`createElement {"class": "test", "content": "test content", "element": "span"} 1`] = `
<span
  class="test"
>
  test content
</span>
`;

exports[`createElement {"class": "test", "element": "a"} 1`] = `
<a
  class="test"
/>
`;

exports[`createElement {"class": "test", "element": "div"} 1`] = `
<div
  class="test"
/>
`;

exports[`createElement {"class": "test", "element": "input"} 1`] = `
<input
  class="test"
/>
`;

exports[`createElement {"class": "test", "element": "span"} 1`] = `
<span
  class="test"
/>
`;

exports[`createElement {"class": [Array], "content": "test content", "element": "a"} 1`] = `
<a
  class="test1 test2 test3"
>
  test content
</a>
`;

exports[`createElement {"class": [Array], "content": "test content", "element": "div"} 1`] = `
<div
  class="test1 test2 test3"
>
  test content
</div>
`;

exports[`createElement {"class": [Array], "content": "test content", "element": "span"} 1`] = `
<span
  class="test1 test2 test3"
>
  test content
</span>
`;

exports[`createElement {"class": [Array], "element": "input"} 1`] = `
<input
  class="test1 test2 test3"
/>
`;

exports[`createElement {"element": "a"} 1`] = `<a />`;

exports[`createElement {"element": "div"} 1`] = `<div />`;

exports[`createElement {"element": "input"} 1`] = `<input />`;

exports[`createElement {"element": "span"} 1`] = `<span />`;
